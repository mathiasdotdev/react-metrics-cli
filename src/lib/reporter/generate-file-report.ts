import {
  AnalysisResult,
  Declaration,
  DeclarationType,
  isUsed,
} from '$types/analysis';
import * as fs from 'fs/promises';
import * as path from 'path';

/**
 * Construit le contenu du rapport d√©taill√©
 */
export const buildDetailedReport = (result: AnalysisResult): string => {
  const lines: string[] = [];

  // En-t√™te
  lines.push("=== RAPPORT D'ANALYSE REACT-METRICS ===");
  lines.push(`G√©n√©r√© le: ${new Date().toLocaleString('fr-FR')}\n`);

  // Statistiques g√©n√©rales
  addStatistics(lines, result);

  // Toutes les d√©clarations trouv√©es
  addAllDeclarations(lines, result);

  // Code mort d√©tect√©
  addDeadCode(lines, result);

  // D√©clarations @deprecated
  addDeprecatedDeclarations(lines, result);

  return lines.join('\n');
};

/**
 * Ajoute les statistiques g√©n√©rales
 */
const addStatistics = (lines: string[], result: AnalysisResult): void => {
  lines.push('=== STATISTIQUES ===');

  const totalDeclarations = result.declarations.size;
  const deadCode = result.deadCode;
  const totalDeadCode = deadCode.length;

  lines.push(`Total d√©clarations trouv√©es: ${totalDeclarations}`);
  lines.push(`Code mort d√©tect√©: ${totalDeadCode}`);

  if (totalDeclarations > 0) {
    const percentage = (totalDeadCode / totalDeclarations) * 100;
    lines.push(`Pourcentage de code mort: ${percentage.toFixed(2)}%`);
  }

  // Statistiques par type
  addStatisticsByType(lines, result);

  // Statistiques @deprecated
  if (result.deprecated && result.deprecated.length > 0) {
    lines.push(`\nD√©clarations @deprecated: ${result.deprecated.length}`);
  }

  lines.push('');
};

/**
 * Ajoute les statistiques par type
 */
const addStatisticsByType = (lines: string[], result: AnalysisResult): void => {
  const counters = new Map<DeclarationType, number>();
  const deadCounters = new Map<DeclarationType, number>();

  // Compter toutes les d√©clarations
  for (const declaration of result.declarations.values()) {
    counters.set(declaration.type, (counters.get(declaration.type) || 0) + 1);
    if (!isUsed(declaration)) {
      deadCounters.set(
        declaration.type,
        (deadCounters.get(declaration.type) || 0) + 1,
      );
    }
  }

  lines.push('\nR√©partition par type:');

  const types = [
    DeclarationType.CONSTANT,
    DeclarationType.FUNCTION,
    DeclarationType.CLASS,
    DeclarationType.CONSOLE,
    DeclarationType.PROP,
    DeclarationType.DEFINITION,
    DeclarationType.EXPORT,
    DeclarationType.DEPENDENCY,
  ];

  for (const type of types) {
    const total = counters.get(type) || 0;
    const dead = deadCounters.get(type) || 0;
    if (total > 0) {
      const percentage = (dead / total) * 100;
      lines.push(
        `  ${type}: ${dead}/${total} (${percentage.toFixed(1)}% mort)`,
      );
    }
  }
};

/**
 * Ajoute toutes les d√©clarations trouv√©es
 */
const addAllDeclarations = (lines: string[], result: AnalysisResult): void => {
  lines.push('=== TOUTES LES D√âCLARATIONS ===');

  // Convertir en array et trier
  const declarations = Array.from(result.declarations.values());
  declarations.sort((a, b) => {
    if (a.type !== b.type) {
      return a.type.localeCompare(b.type);
    }
    return a.name.localeCompare(b.name);
  });

  let currentType: DeclarationType | null = null;
  for (const declaration of declarations) {
    if (declaration.type !== currentType) {
      currentType = declaration.type;
      lines.push(`\n--- ${currentType} ---`);
    }

    let status = 'UTILIS√âE';
    if (!isUsed(declaration)) {
      status = 'NON UTILIS√âE';
    } else if (
      declaration.type === DeclarationType.EXPORT &&
      declaration.isUsedLocally &&
      !declaration.isImportedExternally
    ) {
      status = 'UTILIS√âE LOCALEMENT (supprimer export)';
    }

    lines.push(
      `${declaration.name} | ${declaration.location.filePath}:${declaration.location.line}:${declaration.location.column} | ${status}`,
    );
  }

  lines.push('');
};

/**
 * Ajoute la section du code mort
 */
const addDeadCode = (lines: string[], result: AnalysisResult): void => {
  lines.push('=== CODE MORT D√âTECT√â ===');

  const deadCode = result.deadCode;
  if (deadCode.length === 0) {
    lines.push('Aucun code mort d√©tect√© ! üéâ');
    return;
  }

  // Grouper par type
  const groups = groupByType(deadCode);

  // Afficher chaque groupe dans l'ordre
  const types = [
    DeclarationType.FUNCTION,
    DeclarationType.CLASS,
    DeclarationType.CONSTANT,
    DeclarationType.PROP,
    DeclarationType.CONSOLE,
    DeclarationType.DEFINITION,
    DeclarationType.EXPORT,
    DeclarationType.DEPENDENCY,
  ];

  for (const type of types) {
    const declarations = groups.get(type);
    if (declarations && declarations.length > 0) {
      lines.push(`\n--- ${type} ---`);
      for (const declaration of declarations) {
        lines.push(
          `Fichier: ${declaration.location.filePath}:${declaration.location.line}:${declaration.location.column}`,
        );
        lines.push(`       Nom: ${declaration.name}\n`);
      }
    }
  }
};

/**
 * Groupe les d√©clarations par type
 */
const groupByType = (
  declarations: Declaration[],
): Map<DeclarationType, Declaration[]> => {
  const groups = new Map<DeclarationType, Declaration[]>();

  for (const declaration of declarations) {
    if (!groups.has(declaration.type)) {
      groups.set(declaration.type, []);
    }
    groups.get(declaration.type)!.push(declaration);
  }

  // Trier chaque groupe par nom
  for (const group of groups.values()) {
    group.sort((a, b) => a.name.localeCompare(b.name));
  }

  return groups;
};

/**
 * Ajoute la section des d√©clarations @deprecated
 */
const addDeprecatedDeclarations = (
  lines: string[],
  result: AnalysisResult,
): void => {
  lines.push('\n=== D√âCLARATIONS @DEPRECATED ===');

  const deprecated = result.deprecated;
  if (!deprecated || deprecated.length === 0) {
    lines.push('Aucune d√©claration @deprecated d√©tect√©e.');
    return;
  }

  lines.push(
    'Suggestion: Consid√©rer la suppression ou le remplacement de ces d√©clarations.\n',
  );

  // Grouper par type
  const groups = groupByType(deprecated);

  // Afficher chaque groupe dans l'ordre
  const types = [
    DeclarationType.FUNCTION,
    DeclarationType.CLASS,
    DeclarationType.CONSTANT,
    DeclarationType.PROP,
    DeclarationType.DEFINITION,
    DeclarationType.EXPORT,
  ];

  for (const type of types) {
    const declarations = groups.get(type);
    if (declarations && declarations.length > 0) {
      lines.push(`\n--- ${type} ---`);
      for (const declaration of declarations) {
        lines.push(
          `Fichier: ${declaration.location.filePath}:${declaration.location.line}:${declaration.location.column}`,
        );
        lines.push(`       Nom: ${declaration.name}`);
        lines.push(
          `       Statut: ${isUsed(declaration) ? 'UTILIS√âE' : 'NON UTILIS√âE'}\n`,
        );
      }
    }
  }
};

/**
 * √âcrit du contenu dans un fichier
 * Cr√©e automatiquement les dossiers parents si n√©cessaire
 */
const writeFile = async (filePath: string, content: string): Promise<void> => {
  try {
    const dir = path.dirname(filePath);
    await fs.mkdir(dir, { recursive: true });
    await fs.writeFile(filePath, content, 'utf-8');
  } catch (error) {
    throw new Error(`Impossible d'√©crire le fichier ${filePath}: ${error}`);
  }
};

/**
 * G√©n√®re un rapport d√©taill√© dans un fichier
 */
export const generateFileReport = async (
  filePath: string,
  result: AnalysisResult,
): Promise<void> => {
  const content = buildDetailedReport(result);
  await writeFile(filePath, content);
};
