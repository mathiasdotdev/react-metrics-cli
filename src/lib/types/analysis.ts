/**
 * Type de déclaration détectée dans le code
 */
export enum DeclarationType {
  CONSTANT = 'CONSTANT',
  FUNCTION = 'FUNCTION',
  CLASS = 'CLASS',
  CONSOLE = 'CONSOLE',
  PROP = 'PROP',
  EXPORT = 'EXPORT',
  DEPENDENCY = 'DEPENDENCY',
  DEFINITION = 'DEFINITION',
}

/**
 * Localisation d'une déclaration dans un fichier
 */
export interface Location {
  filePath: string;
  line: number;
  column: number;
}

/**
 * Représente une déclaration trouvée dans le code
 */
export interface Declaration {
  name: string;
  type: DeclarationType;
  location: Location;
  isUsedLocally: boolean;
  isImportedExternally: boolean;
  isDeprecated?: boolean;
  context?: string;
  metadata?: Record<string, unknown>;
}

/**
 * Résultat complet de l'analyse
 */
export interface AnalysisResult {
  declarations: Map<string, Declaration>;
  deadCode: Declaration[];
  deprecated: Declaration[];
  duration: number;
}

/**
 * Génère une clé unique pour une déclaration
 */
export const getDeclarationKey = (declaration: Declaration): string => {
  return `${declaration.location.filePath}:${declaration.location.line}:${declaration.location.column}:${declaration.name}`;
};

/**
 * Vérifie si une déclaration est utilisée (localement ou importée ailleurs)
 */
export const isUsed = (declaration: Declaration): boolean => {
  return declaration.isUsedLocally || declaration.isImportedExternally;
};

/**
 * Récupère le code mort (déclarations non utilisées)
 */
export const getDeadCode = (
  declarations: Map<string, Declaration>,
): Declaration[] => {
  return Array.from(declarations.values()).filter((d) => !isUsed(d));
};

/**
 * Récupère toutes les déclarations marquées @deprecated
 */
export const getDeprecated = (
  declarations: Map<string, Declaration>,
): Declaration[] => {
  return Array.from(declarations.values())
    .filter((d) => d.isDeprecated === true)
    .sort((a, b) => a.location.filePath.localeCompare(b.location.filePath));
};
