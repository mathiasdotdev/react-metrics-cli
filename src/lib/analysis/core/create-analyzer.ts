/**
 * Analyseur principal qui coordonne la détection et la vérification
 */

import { AnalysisResult, getDeadCode } from '$types/analysis';
import { DetectorConfig } from '$types/config';
import { createDetectionCoordinator } from '@detection/create-detection-coordinator';
import { createVerificationCoordinator } from '@verification/create-verification-coordinator';

/**
 * Crée un analyseur de code mort
 */
export const createAnalyzer = (config: DetectorConfig, projectPath: string) => {
  // Créer les coordinateurs
  const detectionCoordinator = createDetectionCoordinator(config, projectPath);
  const verificationCoordinator = createVerificationCoordinator(config);

  return {
    /**
     * Exécute l'analyse complète (détection + vérification)
     */
    executeCompleteAnalysis: async (): Promise<AnalysisResult> => {
      const startTime = Date.now();

      if (config.debug) {
        console.log("\n[DEBUG] === DÉBUT DE L'ANALYSE ===\n");
      }

      // Phase 1 : Détection des déclarations
      if (config.debug) {
        console.log('[DEBUG] Phase 1: Détection des déclarations...\n');
      }
      const result = await detectionCoordinator.executeDetection();

      // Phase 2 : Vérification des utilisations
      await verificationCoordinator.executeVerification(result);

      // Finaliser : récupérer le code mort
      result.deadCode = getDeadCode(result.declarations);
      result.duration = Date.now() - startTime;

      return result;
    },
  };
};
