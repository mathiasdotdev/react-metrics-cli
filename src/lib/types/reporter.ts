import { AnalysisResult } from './analysis';

/**
 * Interface abstraite pour les reporters
 */
export interface Reporter {
  /**
   * Génère un rapport à partir des résultats
   */
  generateReport(result: AnalysisResult): void | Promise<void>;
}
