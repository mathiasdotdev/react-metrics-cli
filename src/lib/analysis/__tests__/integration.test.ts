import type { AnalysisResult } from '$types/analysis';
import { DeclarationType } from '$types/analysis';
import { createAnalyzer } from '@core/create-analyzer';
import { buildDetailedReport } from '$lib/reporter/generate-file-report';
import path from 'node:path';
import { beforeAll, describe, expect, it } from 'bun:test';

describe('Integration Test - React Demo Analysis', () => {
  let result: AnalysisResult;
  const projectPath = path.resolve(process.cwd(), 'react-demo');

  beforeAll(async () => {
    const config = {
      fileExtensions: ['.js', '.jsx', '.ts', '.tsx'],
      ignoredFolders: ['node_modules', '.git', 'dist', 'build', 'coverage'],
      ignoreAnnotations: true,
      ignoreFileAnnotation: 'react-metrics-ignore-file',
      ignoreLineAnnotation: 'react-metrics-ignore-ligne',
      contextLines: 5,
    };

    const analyzer = createAnalyzer(config, projectPath);
    result = await analyzer.executeCompleteAnalysis();
  });

  describe('Total Dead Code Detection', () => {
    it('should detect exactly 87 dead code elements', () => {
      expect(result.deadCode.length).toBe(87);
    });

    it('should have valid duration', () => {
      expect(result.duration).toBeGreaterThan(0);
      expect(result.duration).toBeLessThan(5000);
    });

    it('should match expected counts by type', () => {
      const counts = {
        functions: result.deadCode.filter(
          (d) => d.type === DeclarationType.FUNCTION,
        ).length,
        classes: result.deadCode.filter((d) => d.type === DeclarationType.CLASS)
          .length,
        constants: result.deadCode.filter(
          (d) => d.type === DeclarationType.CONSTANT,
        ).length,
        props: result.deadCode.filter((d) => d.type === DeclarationType.PROP)
          .length,
        consoles: result.deadCode.filter(
          (d) => d.type === DeclarationType.CONSOLE,
        ).length,
        definitions: result.deadCode.filter(
          (d) => d.type === DeclarationType.DEFINITION,
        ).length,
        exports: result.deadCode.filter(
          (d) => d.type === DeclarationType.EXPORT,
        ).length,
        dependencies: result.deadCode.filter(
          (d) => d.type === DeclarationType.DEPENDENCY,
        ).length,
      };

      expect(counts).toEqual({
        functions: 11,
        classes: 4,
        constants: 6,
        props: 13,
        consoles: 13,
        definitions: 8,
        exports: 27,
        dependencies: 5,
      });
    });
  });

  describe('Functions Detection Details', () => {
    it('should detect all expected function names', () => {
      const functionNames = result.deadCode
        .filter((d) => d.type === DeclarationType.FUNCTION)
        .map((d) => d.name)
        .sort();

      expect(functionNames).toMatchSnapshot();
    });

    it('should detect specific function details', () => {
      const functionsDetails = result.deadCode
        .filter((d) => d.type === DeclarationType.FUNCTION)
        .map((d) => ({
          name: d.name,
          file: d.location.filePath.replace(/\\/g, '/').split('react-demo/')[1],
          line: d.location.line,
          column: d.location.column,
        }))
        .sort((a, b) => a.name.localeCompare(b.name));

      expect(functionsDetails).toMatchSnapshot();
    });
  });

  describe('Classes Detection Details', () => {
    it('should detect 4 unused classes', () => {
      const classes = result.deadCode.filter(
        (d) => d.type === DeclarationType.CLASS,
      );
      expect(classes.length).toBe(4);
    });

    it('should detect all class names and locations', () => {
      const classDetails = result.deadCode
        .filter((d) => d.type === DeclarationType.CLASS)
        .map((d) => ({
          name: d.name,
          file: d.location.filePath.replace(/\\/g, '/').split('react-demo/')[1],
          line: d.location.line,
        }))
        .sort(
          (a, b) =>
            a.name.localeCompare(b.name) || a.file.localeCompare(b.file),
        );

      expect(classDetails).toMatchSnapshot();
    });
  });

  describe('Constants Detection Details', () => {
    it('should detect 6 unused constants', () => {
      const constants = result.deadCode.filter(
        (d) => d.type === DeclarationType.CONSTANT,
      );
      expect(constants.length).toBe(6);
    });

    it('should detect all constant names', () => {
      const constantNames = result.deadCode
        .filter((d) => d.type === DeclarationType.CONSTANT)
        .map((d) => d.name)
        .sort();

      expect(constantNames).toMatchSnapshot();
    });
  });

  describe('Props Detection Details', () => {
    it('should detect 13 unused props', () => {
      const props = result.deadCode.filter(
        (d) => d.type === DeclarationType.PROP,
      );
      expect(props.length).toBe(13);
    });

    it('should detect all prop names', () => {
      const propNames = result.deadCode
        .filter((d) => d.type === DeclarationType.PROP)
        .map((d) => d.name)
        .sort();

      expect(propNames).toMatchSnapshot();
    });
  });

  describe('Console Logs Detection Details', () => {
    it('should detect 13 console statements', () => {
      const consoles = result.deadCode.filter(
        (d) => d.type === DeclarationType.CONSOLE,
      );
      expect(consoles.length).toBe(13);
    });

    it('should detect console types distribution', () => {
      const consoleTypes = result.deadCode
        .filter((d) => d.type === DeclarationType.CONSOLE)
        .map((d) => d.context)
        .filter((ctx): ctx is string => ctx !== undefined)
        .sort();

      expect(consoleTypes).toMatchSnapshot();
    });
  });

  describe('Types/Interfaces Detection Details', () => {
    it('should detect 8 unused types/interfaces', () => {
      const definitions = result.deadCode.filter(
        (d) => d.type === DeclarationType.DEFINITION,
      );
      expect(definitions.length).toBe(8);
    });

    it('should detect all type/interface names', () => {
      const defNames = result.deadCode
        .filter((d) => d.type === DeclarationType.DEFINITION)
        .map((d) => d.name)
        .sort();

      expect(defNames).toMatchSnapshot();
    });
  });

  describe('Exports Detection Details', () => {
    it('should detect 27 unused exports', () => {
      const exports = result.deadCode.filter(
        (d) => d.type === DeclarationType.EXPORT,
      );
      expect(exports.length).toBe(27);
    });

    it('should detect all export names', () => {
      const exportNames = result.deadCode
        .filter((d) => d.type === DeclarationType.EXPORT)
        .map((d) => d.name)
        .sort();

      expect(exportNames).toMatchSnapshot();
    });
  });

  describe('Dependencies Detection Details', () => {
    it('should detect 5 unused dependencies', () => {
      const deps = result.deadCode.filter(
        (d) => d.type === DeclarationType.DEPENDENCY,
      );
      expect(deps.length).toBe(5);
    });

    it('should detect dependency details', () => {
      const depDetails = result.deadCode
        .filter((d) => d.type === DeclarationType.DEPENDENCY)
        .map((d) => ({
          name: d.name,
          file: d.location.filePath.replace(/\\/g, '/').split('react-demo/')[1],
        }));

      expect(depDetails).toMatchSnapshot();
    });
  });

  describe('Debug Mode Coverage', () => {
    it('should execute analysis in debug mode and log verification details', async () => {
      const debugConfig = {
        fileExtensions: ['.js', '.jsx', '.ts', '.tsx'],
        ignoredFolders: ['node_modules', '.git', 'dist', 'build', 'coverage'],
        ignoreAnnotations: true,
        ignoreFileAnnotation: 'react-metrics-ignore-file',
        ignoreLineAnnotation: 'react-metrics-ignore-ligne',
        contextLines: 5,
        debug: true, // Activer le mode debug
      };

      // Capturer les logs console
      const originalLog = console.log;
      const logs: string[] = [];
      console.log = (...args: any[]) => {
        logs.push(args.join(' '));
      };

      const debugAnalyzer = createAnalyzer(debugConfig, projectPath);
      const debugResult = await debugAnalyzer.executeCompleteAnalysis();

      // Restaurer console.log
      console.log = originalLog;

      // Vérifier que l'analyse a réussi
      expect(debugResult.deadCode.length).toBeGreaterThan(0);

      // Vérifier que des logs debug ont été générés
      expect(logs.length).toBeGreaterThan(0);

      // Vérifier qu'on a des logs de vérification
      const verificationLogs = logs.filter((log) =>
        log.includes('[DEBUG VERIFICATION]'),
      );
      expect(verificationLogs.length).toBeGreaterThan(0);

      // Vérifier qu'on a le résumé
      const summaryLogs = logs.filter((log) => log.includes('RÉSUMÉ'));
      expect(summaryLogs.length).toBeGreaterThan(0);
    }, 10000);

    it('should handle file read errors gracefully', async () => {
      const config = {
        fileExtensions: ['.js', '.jsx', '.ts', '.tsx'],
        ignoredFolders: ['node_modules', '.git', 'dist', 'build', 'coverage'],
        ignoreAnnotations: true,
        ignoreFileAnnotation: 'react-metrics-ignore-file',
        ignoreLineAnnotation: 'react-metrics-ignore-ligne',
        contextLines: 5,
        debug: false,
      };

      // Capturer les warnings console
      const originalWarn = console.warn;
      const warnings: string[] = [];
      console.warn = (...args: any[]) => {
        warnings.push(args.join(' '));
      };

      const analyzer = createAnalyzer(config, projectPath);
      const testResult = await analyzer.executeCompleteAnalysis();

      // Restaurer console.warn
      console.warn = originalWarn;

      // L'analyse doit réussir même si certains fichiers ne peuvent pas être lus
      expect(testResult).toBeDefined();
      expect(testResult.declarations.size).toBeGreaterThan(0);
    }, 10000);
  });

  describe('Report Generation', () => {
    it('should generate complete report for react-demo project', () => {
      // Réutiliser le result calculé dans beforeAll
      const report = buildDetailedReport(result);

      // Vérifications structurelles du rapport
      expect(report).toContain("=== RAPPORT D'ANALYSE REACT-METRICS ===");
      expect(report).toContain('=== STATISTIQUES ===');
      expect(report).toContain('=== TOUTES LES DÉCLARATIONS ===');
      expect(report).toContain('=== CODE MORT DÉTECTÉ ===');
      expect(report).toContain('=== DÉCLARATIONS @DEPRECATED ===');

      // Vérifications de contenu (basées sur react-demo)
      expect(report).toContain('Total déclarations trouvées:');
      expect(report).toContain('Code mort détecté:');
      expect(report).toContain('Pourcentage de code mort:');
      expect(report).toContain('Répartition par type:');

      // Vérifier que les @deprecated sont présents (DeprecatedFeatures.ts)
      expect(report).toMatch(/Déclarations @deprecated:\s+\d+/);

      // Vérifier structure des sections par type
      expect(report).toMatch(/--- FUNCTION ---/);
      expect(report).toMatch(/--- CLASS ---/);
      expect(report).toMatch(/--- CONSTANT ---/);
    });
  });
});
