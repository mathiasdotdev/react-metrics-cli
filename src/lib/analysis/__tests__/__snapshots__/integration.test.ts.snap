// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Integration Test - React Demo Analysis Functions Detection Details should detect all expected function names 1`] = `
[
  "Divide2Numbers",
  "FunctionButton",
  "appFetch",
  "divide",
  "oldArrowFunction",
  "oldFunction",
  "processUser",
  "specificIgnoredFunction",
  "temporaryFunction",
  "thisWillAlsoBeDetected",
  "unusedFunction",
]
`;

exports[`Integration Test - React Demo Analysis Functions Detection Details should detect specific function details 1`] = `
[
  {
    "column": 16,
    "file": "src/config/fetch.ts",
    "line": 7,
    "name": "appFetch",
  },
  {
    "column": 10,
    "file": "src/features/cas1.ts",
    "line": 27,
    "name": "divide",
  },
  {
    "column": 10,
    "file": "src/features/cas2.ts",
    "line": 20,
    "name": "Divide2Numbers",
  },
  {
    "column": 10,
    "file": "src/features/advanced-props.tsx",
    "line": 84,
    "name": "FunctionButton",
  },
  {
    "column": 14,
    "file": "src/features/DeprecatedFeatures.ts",
    "line": 42,
    "name": "oldArrowFunction",
  },
  {
    "column": 17,
    "file": "src/features/DeprecatedFeatures.ts",
    "line": 9,
    "name": "oldFunction",
  },
  {
    "column": 10,
    "file": "src/types/TestDefinitions.ts",
    "line": 72,
    "name": "processUser",
  },
  {
    "column": 10,
    "file": "src/features/annotations.tsx",
    "line": 19,
    "name": "specificIgnoredFunction",
  },
  {
    "column": 10,
    "file": "src/features/annotations.tsx",
    "line": 34,
    "name": "temporaryFunction",
  },
  {
    "column": 10,
    "file": "src/features/ignore-file-late-comment.jsx",
    "line": 6,
    "name": "thisWillAlsoBeDetected",
  },
  {
    "column": 7,
    "file": "src/components/Button.tsx",
    "line": 31,
    "name": "unusedFunction",
  },
]
`;

exports[`Integration Test - React Demo Analysis Classes Detection Details should detect all class names and locations 1`] = `
[
  {
    "file": "src/features/DeprecatedFeatures.ts",
    "line": 16,
    "name": "OldClass",
  },
  {
    "file": "src/components/ClassExample.tsx",
    "line": 11,
    "name": "UnusedClass",
  },
  {
    "file": "src/features/cas1.ts",
    "line": 32,
    "name": "UnusedClass",
  },
  {
    "file": "src/features/cas2.ts",
    "line": 25,
    "name": "UnusedClass",
  },
]
`;

exports[`Integration Test - React Demo Analysis Constants Detection Details should detect all constant names 1`] = `
[
  "OLD_CONSTANT",
  "blabla2",
  "normalConstant",
  "thisVariableWillAlsoBeDetected",
  "thisVariableWillBeDetected",
  "toto",
]
`;

exports[`Integration Test - React Demo Analysis Props Detection Details should detect all prop names 1`] = `
[
  "active",
  "appStatus",
  "counter",
  "data",
  "errorElement",
  "metadata",
  "newProp",
  "promise",
  "signal",
  "total",
  "unused",
  "user",
  "users",
]
`;

exports[`Integration Test - React Demo Analysis Console Logs Detection Details should detect console types distribution 1`] = `
[
  "console.debug",
  "console.error",
  "console.info",
  "console.log",
  "console.log",
  "console.log",
  "console.log",
  "console.log",
  "console.log",
  "console.log",
  "console.log",
  "console.log",
  "console.warn",
]
`;

exports[`Integration Test - React Demo Analysis Types/Interfaces Detection Details should detect all type/interface names 1`] = `
[
  "CombinedStatus",
  "DeprecatedProps",
  "OldInterface",
  "OldType",
  "UnusedExportedInterface",
  "UnusedExportedType",
  "UnusedNotExportedInterface",
  "UnusedNotExportedType",
]
`;

exports[`Integration Test - React Demo Analysis Exports Detection Details should detect all export names 1`] = `
[
  "App",
  "AuthenticatedButton",
  "BasicButton",
  "ButtonSimplified",
  "ClassExample",
  "ComplexButton",
  "DefaultButton",
  "ForwardRefButton",
  "MemoizedButton",
  "MultiLineButton",
  "NEW_CONSTANT",
  "NewClass",
  "ROUTE_ALL",
  "ROUTE_DASHBOARD",
  "ROUTE_DEFAULT",
  "ROUTE_LOGIN",
  "ROUTE_SIGNUP",
  "RenamedButton",
  "SpreadButton",
  "Subtract2Numbers",
  "UnusedButton",
  "UnusedExportedClass",
  "UnusedExportedClass",
  "goodbye",
  "newFunction",
  "router",
  "withAuth",
]
`;

exports[`Integration Test - React Demo Analysis Dependencies Detection Details should detect dependency details 1`] = `
[
  {
    "file": "package.json",
    "name": "@types/react-router-dom",
  },
  {
    "file": "package.json",
    "name": "react-dom",
  },
  {
    "file": "package.json",
    "name": "@types/react",
  },
  {
    "file": "package.json",
    "name": "@types/react-dom",
  },
  {
    "file": "package.json",
    "name": "typescript",
  },
]
`;
